x/rollapp/transfergenesis/ibc_module.go

```diff
@@ -0,0 +1,260 @@
+
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/transfersenabled"
+
+	"github.com/dymensionxyz/dymension/v3/utils/derr"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	uibc "github.com/dymensionxyz/dymension/v3/utils/ibc"
+
+	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+
+	"github.com/dymensionxyz/dymension/v3/utils/gerr"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	transfertypes "github.com/cosmos/ibc-go/v6/modules/apps/transfer/types"
+
+	errorsmod "cosmossdk.io/errors"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+	delayedackkeeper "github.com/dymensionxyz/dymension/v3/x/delayedack/keeper"
+	rollappkeeper "github.com/dymensionxyz/dymension/v3/x/rollapp/keeper"
+	rollapptypes "github.com/dymensionxyz/dymension/v3/x/rollapp/types"
+)
+
+const (
+	memoNamespaceKey = "genesis_transfer"
+)
+
+type DenomMetadataKeeper interface {
+	CreateDenomMetadata(ctx sdk.Context, metadata banktypes.Metadata) error
+	HasDenomMetadata(ctx sdk.Context, base string) bool
+}
+
+type TransferKeeper interface {
+	SetDenomTrace(ctx sdk.Context, denomTrace transfertypes.DenomTrace)
+}
+
+type IBCModule struct {
+	porttypes.IBCModule // next one
+	delayedackKeeper    delayedackkeeper.Keeper
+	rollappKeeper       rollappkeeper.Keeper
+	transferKeeper      TransferKeeper
+	denomKeeper         DenomMetadataKeeper
+}
+
+func NewIBCModule(
+	next porttypes.IBCModule,
+	delayedAckKeeper delayedackkeeper.Keeper,
+	rollappKeeper rollappkeeper.Keeper,
+	transferKeeper TransferKeeper,
+	denomKeeper DenomMetadataKeeper,
+) IBCModule {
+	return IBCModule{
+		IBCModule:        next,
+		delayedackKeeper: delayedAckKeeper,
+		rollappKeeper:    rollappKeeper,
+		transferKeeper:   transferKeeper,
+		denomKeeper:      denomKeeper,
+	}
+}
+
+func (w IBCModule) logger(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+) log.Logger {
+	return ctx.Logger().With(
+		"module", "transferGenesis",
+		"packet_source_port", packet.SourcePort,
+		"packet_destination_port", packet.DestinationPort,
+		"packet_sequence", packet.Sequence,
+		"method", "OnRecvPacket",
+	)
+}
+
+// OnRecvPacket will, if the packet is a transfer packet:
+// if it's not a genesis transfer: pass on the packet only if transfers are enabled
+// else: check it's a valid genesis transfer. If it is, then register the denom, if
+// it's the last one, open the bridge.
+// NOTE: we assume that by this point the canonical channel ID has already been set
+// for the rollapp, in a secure way.
+func (w IBCModule) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	l := w.logger(ctx, packet)
+
+	if !w.delayedackKeeper.IsRollappsEnabled(ctx) {
+		l.Debug("Rollapps are disabled.")
```

Comment:
those checks duplicated along the middlewares. no need to log in each layer

Reply:
removed